### PR TITLE
Add RAG model and utilities for Mindsdb integration

### PR DESCRIPTION
**Pull request by dusvyat**
## Description

Porting over RAG improvements from https://github.com/mindsdb/email_rag

TBC - we need to decide and best integration point for user. This will be informed by https://linear.app/mindsdb/issue/RAG-3/unify-rag-implementation

Fixes RAG-57

Fixes #issue_number

## Type of change

(Please delete options that are not relevant)

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [ ] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] 📄 This change requires a documentation update

## Verification Process

To ensure the changes are working as expected:

 - [ ]   Test Location: Specify the URL or path for testing.
 - [ ]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.

## Additional Media:

- [ ] I have attached a brief loom video or screenshots showcasing the new functionality or change.

## Checklist:

- [ ] My code follows the style guidelines(PEP 8) of MindsDB.
- [ ] I have appropriately commented on my code, especially in complex areas.
- [ ] Necessary documentation updates are either made or tracked in issues.
- [ ] Relevant unit and integration tests are updated or added.

### CODE REVIEW

**Comment by dusvyat** [comment]
cc @QuantumPlumber - starting to integrate rag improvements here

File: mindsdb/integrations/utilities/rag/settings.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    persist_directory: str = None
+    collection_name: str = None
+    connection_string: str = None
+
+    class Config:
+        arbitrary_types_allowed = True
+        extra = "forbid"
+
+
+class RAGPipelineModel(BaseModel):
```
Comment (dusvyat) [inline mindsdb/integrations/utilities/rag/settings.py:141]: @ea-rus this is config for rag

File: tests/handler_tests/test_rag_pipelines.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    config['documents'] = create_test_documents()
+    config['llm'] = DEFAULT_LLM
+    config['embeddings_model'] = DEFAULT_EMBEDDINGS
+
+    return config
+
+
+def test_rag_pipeline_creation(config):
+    rag = RAG(config)
+    result = rag.pipeline.invoke('test document')
```
Comment (dusvyat) [inline tests/handler_tests/test_rag_pipelines.py:44]: @ea-rus an example of how we would run RAG

File: mindsdb/integrations/utilities/rag/rag_pipeline_builder.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+def get_pipeline_from_retriever(config: RAGPipelineModel) -> RunnableSerializable:
+    retriever_strategy = _retriever_strategies.get(config.retriever_type)
+    if retriever_strategy:
+        return retriever_strategy(config).with_returned_sources()
+    else:
+        raise ValueError(
+            f'Invalid retriever type, must be one of: {list(_retriever_strategies.keys())}. Got {config.retriever_type}')
+
+
+class RAG:
```
Comment (dusvyat) [inline mindsdb/integrations/utilities/rag/rag_pipeline_builder.py:80]: @ea-rus in this module we construct the rag pipeline

File: mindsdb/integrations/utilities/rag/loaders/vector_store_loader/vector_store_loader.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+        if settings.vector_store_type == VectorStoreType.CHROMA:
+            return VectorStoreFactory._load_chromadb_store(embeddings_model, settings)
+        elif settings.vector_store_type == VectorStoreType.PGVECTOR:
+            return VectorStoreFactory._load_pgvector_store(embeddings_model, settings)
+        else:
+            raise ValueError(f"Invalid vector store type, must be one either {VectorStoreType.__members__.keys()}")
+
+    @staticmethod
+    def _load_chromadb_store(embeddings_model: Embeddings, settings) -> Chroma:
```
Comment (ea-rus) [inline mindsdb/integrations/utilities/rag/loaders/vector_store_loader/vector_store_loader.py:55]: this method (looking at the name) should be similar to `_load_pgvector_store`
why it doesn't use `_load_data_into_langchain_pgvector` ?

File: mindsdb/integrations/utilities/rag/loaders/vector_store_loader/vector_store_loader.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            texts=texts,
+            embeddings=embeddings,
+            metadatas=metadata,
+            ids=ids
+        )
+        return vectorstore
+
+    @staticmethod
+    def _fetch_data_from_db(settings: VectorStoreConfig) -> pd.DataFrame:
+        """
```
Comment (ea-rus) [inline mindsdb/integrations/utilities/rag/loaders/vector_store_loader/vector_store_loader.py:101]: should the data extraction be outside of vectorstore loader module?
besides db we could have pdf/url source and maybe data could be extracted somewhere in retrievers and  be passed to vectorstore loader as dataframe

File: mindsdb/integrations/utilities/rag/loaders/vector_store_loader/vector_store_loader.py (hunk 4 above)
Comment (dusvyat) [reply mindsdb/integrations/utilities/rag/loaders/vector_store_loader/vector_store_loader.py:55]: loading an existing chromaDB instance works for all chromaDB even if created outside Langchain.

However, for PGVector if it is created outside of Langchain (i.e Mindsdb) we need to map data to langchain schema. see `_fetch_data_from_db`

File: mindsdb/integrations/utilities/rag/loaders/vector_store_loader/vector_store_loader.py (hunk 5 above)
Comment (dusvyat) [reply mindsdb/integrations/utilities/rag/loaders/vector_store_loader/vector_store_loader.py:101]: This is used to load existing pgvector db that was created outside of langchain

File: mindsdb/integrations/utilities/rag/callback_handlers/log_callback_handler.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
+from typing import Any, Dict, List, Union
```
[the remark this answers is not in the document]
Comment (dusvyat) [reply mindsdb/integrations/utilities/rag/callback_handlers/log_callback_handler.py:1]: ah yes good point, I will do that

File: mindsdb/integrations/utilities/rag/pipelines/rag.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        vector_store_operator = VectorStoreOperator(
+            vector_store=config.vector_store,
+            documents=config.documents,
+            embeddings_model=config.embeddings_model
+        )
+
+        return cls(vector_store_operator.vector_store.as_retriever(), config.rag_prompt_template, config.llm)
+
+    @classmethod
+    def from_sql_retriever(cls, config: RAGPipelineModel):
```
[the remark this answers is not in the document]
Comment (dusvyat) [reply mindsdb/integrations/utilities/rag/pipelines/rag.py]: @tmichaeldb DMing you thoughts on this

**Pull request merged**